sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1091 @@
+            raise e
+
+    async def _handle_auto_reconnect(self):
+        _LOGGER.debug("start auto reconnect")
+        success = False
+        attempt = 0
+        while not self._is_stopping:
+            try:
+                await self._start_from_restarting()
+                success = True
+                break
+            except Exception as e:  # pylint: disable=broad-except
+                _LOGGER.warning("An attempt to reconnect connection failed %s", e)
+                attempt = attempt + 1
+                delay_seconds = self._reconnect_retry_policy.next_retry_delay(attempt)
+                if not delay_seconds:
+                    break
+                _LOGGER.debug("Delay time for reconnect attempt %d: %ds", attempt, delay_seconds)
+                await asyncio.sleep(delay_seconds)
+        if not success:
+            await self._handle_connection_stopped()
+        else:
+            _LOGGER.debug("reconnect successfully")
+
+    async def _handle_connection_stopped(self):
+        _LOGGER.debug("Connection stopped")
+        self._is_stopping = False
+        self._state = WebPubSubClientState.STOPPED
+        self._call_back(CallbackType.STOPPED)
+
+    async def _handle_connection_close_and_no_recovery(self):
+        _LOGGER.debug("Connection closed and no recovery")
+        self._state = WebPubSubClientState.DISCONNECTED
+        self._call_back(
+            CallbackType.DISCONNECTED,
+            OnDisconnectedArgs(
+                connection_id=self._connection_id,
+                message=self._last_disconnected_message.message if self._last_disconnected_message else None,
+            ),
+        )
+        if self._auto_reconnect:
+            await self._handle_auto_reconnect()
+        else:
+            await self._handle_connection_stopped()
+
+    def _build_recovery_url(self) -> Union[str, None]:
+        if self._connection_id and self._reconnection_token and self._url:
+            params = {
+                "awps_connection_id": self._connection_id,
+                "awps_reconnection_token": self._reconnection_token,
+            }
+            url_parse = urllib.parse.urlparse(self._url)
+            url_dict = dict(urllib.parse.parse_qsl(url_parse.query))
+            url_dict.update(params)
+            new_query = urllib.parse.urlencode(url_dict)
+            url_parse = url_parse._replace(query=new_query)
+            new_url = urllib.parse.urlunparse(url_parse)
+            return str(new_url)
+        return None
+
+    def _is_connected(self) -> bool:
+        """check whether the client is still connected to server after open
+
+        :return: True if the client is connected to server, otherwise False
+        :rtype: bool
+        """
+        return self._state == WebPubSubClientState.CONNECTED and not self._ws.closed  # type: ignore
+
+    async def _rejoin_group(self, group_name: str):
+        async def _rejoin_group():
```

Comment:
is there a reason for an inner method here with the same name as the outer one?

Reply:
No. I change inner name to `_rejoin_group_func` to avoid misunderstanding.